internalshared/listenerutil/tls_acme.go

```diff
@@ -0,0 +1,280 @@
+	Reload() error
+}
+
+type CertGetter interface {
+	GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+type ReloadableCertGetter interface {
+	Reloadable
+	CertGetter
+}
+
+type ACMECertGetter struct {
+	Listener *configutil.Listener
+
+	Magic *certmagic.Config
+	ACME  *certmagic.ACMEIssuer
+}
+
+func NewCertificateGetter(l *configutil.Listener, ui cli.Ui, logger hclog.Logger) (ReloadableCertGetter, error) {
+	// Assume a validated listener here. Prefer the certificate if is set
+	// by path in the listener configuration, otherwise do ACME certificate
+	// acquisition.
+	if l.TLSCertFile != "" {
+		cg := reloadutil.NewCertificateGetter(l.TLSCertFile, l.TLSKeyFile, "")
+		if err := cg.Reload(); err != nil {
+			// We try the key without a passphrase first and if we get an incorrect
+			// passphrase response, try again after prompting for a passphrase
+			if errwrap.Contains(err, x509.IncorrectPasswordError.Error()) {
+				var passphrase string
+				passphrase, err = ui.AskSecret(fmt.Sprintf("Enter passphrase for %s:", l.TLSKeyFile))
+				if err == nil {
+					cg = reloadutil.NewCertificateGetter(l.TLSCertFile, l.TLSKeyFile, passphrase)
+					if err = cg.Reload(); err == nil {
+						return cg, nil
+					}
+
+					return nil, fmt.Errorf("error loading TLS cert with password: %w", err)
+				}
+			}
+
+			return nil, fmt.Errorf("error loading TLS cert: %w", err)
+		}
+
+		return cg, nil
+	}
+
+	core := &zapHclCore{Logger: logger}
+	zapLogger := zap.New(core)
+	if logger == nil {
+		zapLogger = zap.NewNop()
+	}
+	certmagic.Default.Logger = zapLogger
+
+	acg := &ACMECertGetter{
+		Listener: l,
+		Magic:    certmagic.NewDefault(),
+	}
+
+	acg.Magic.OnDemand = new(certmagic.OnDemandConfig)
+	acg.Magic.OnDemand.DecisionFunc = func(ctx context.Context, name string) error {
+		if len(l.TLSACMEDomains) == 0 {
+			return nil
+		}
+
+		found := false
+		for _, allowed := range l.TLSACMEDomains {
+			if strings.EqualFold(allowed, name) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return errors.New("domain not allowed in tls_acme_domains")
+		}
+
+		return nil
+	}
+
+	if l.TLSACMECachePath != "" {
+		acg.Magic.Storage = &certmagic.FileStorage{
+			Path: l.TLSACMECachePath,
+		}
+	}
+
+	if l.TLSACMEKeyType != "" {
+		switch certmagic.KeyType(l.TLSACMEKeyType) {
+		case certmagic.ED25519, certmagic.P256, certmagic.P384, certmagic.RSA2048, certmagic.RSA4096, certmagic.RSA8192:
+		default:
+			return nil, fmt.Errorf("unknown value for tls_acme_key_type (`%v`); allowed values are `%v`, `%v`, `%v`, `%v`, `%v`, `%v`", l.TLSACMEKeyType, certmagic.ED25519, certmagic.P256, certmagic.P384, certmagic.RSA2048, certmagic.RSA4096, certmagic.RSA8192)
+		}
+
+		acg.Magic.KeySource = certmagic.StandardKeyGenerator{
+			KeyType: certmagic.KeyType(l.TLSACMEKeyType),
+		}
+	}
+
+	template := certmagic.ACMEIssuer{
+		CA:                      l.TLSACMECADirectory,
+		TestCA:                  l.TLSACMECADirectory,
```

Comment:
I think we shouldn't set this to the same value as `CA` by default. Either leave it empty or allow the user to set this using a config variable

Reply:
It is not clear from the certmagic docs when this is used. I think it is used when `attempts > 0` perhaps? But that makes no sense.

Reply:
I set this to the empty string. I think that means we won't ever use the TestCA, which is probably ideal behavior, given you can't explicitly tell CertMagic to use the TestCA>

Reply:
By default certmagic sets this to the Let's Encrypt Staging CA
https://github.com/caddyserver/certmagic/blob/v0.21.7/acmeissuer.go#L638-L643

Reply:
@JanMa But only if `CA == DefaultACME.CA`: 

https://github.com/caddyserver/certmagic/blob/3f666293f0e588b804d5bc4e792a841d68a3f27c/acmeissuer.go#L183-L189

Which it doesn't, so it'll be blank?

Reply:
Right, so let's set this blank by default and enable the user to override it if they want?

Reply:
@JanMa Hmm... Like I said, I don't think we ever use it. There's no switch in certmagic to actually force it to use the TestCA, so I'm not sure what value it serves. I think the only case it gets used is if the main issuance fails for some reason, it'll try falling back to the test server to see if that one works, before retrying on the main server.

Mostly I think that's in the interest of avoiding rate limiting in the event of a configuration failure, though CertMagic should already respect the rate limiting (and has its own).

Reply:
Filed as https://github.com/caddyserver/certmagic/issues/330